LL-7025 LL-7059 Language discoverability

tlaforge-ledger opened the pull request:
#### **EDIT: (13/01) @ofreyssinet-ledger:**
new changes explained in this comment: https://github.com/LedgerHQ/ledger-live-mobile/pull/1926#issuecomment-1012268761

TL;DR:
- default initial language (for new users) to system language if that language is supported [LL-7059]
- some refactoring to change where the source of truth for i18next is and where it gets updated
___
#### Initial PR comment:

- use crypto-js to replace react-native-crypto
- add pushed languages
- removed fr from preferred languages
- add checker for language availability
- move languages list
- add translation keys
- add language logo
- checker component
- add component on default screen to use hooks like useTheme

### Type

New Feature

### Context

[LL-7025]
[LL-7059]
[LL-8921]

### Parts of the app affected / Test plan

On startup, on default screen (portfolio), it add a modal first time you start with another language than default on your mobile.
The language must be part of translates trusted list.
It give you the possibility to switch to this language. If you say yes, or not, we handle the answer and remember it so you will no longer get the info.

[LL-7025]: https://ledgerhq.atlassian.net/browse/LL-7025
[LL-7059]: https://ledgerhq.atlassian.net/browse/LL-7059

[LL-8921]: https://ledgerhq.atlassian.net/browse/LL-8921?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

tlaforge-ledger committed:
add pushed languages

tlaforge-ledger committed:
removed fr from preferred languages

tlaforge-ledger committed:
add checker for language availability

tlaforge-ledger committed:
move languages list

tlaforge-ledger committed:
add translation keys

tlaforge-ledger committed:
add language logo

tlaforge-ledger committed:
checker component

tlaforge-ledger committed:
add component on default screen to use hooks like useTheme

tlaforge-ledger committed:
remove some logs

tlaforge-ledger committed:
add track for modal opened

tlaforge-ledger committed:
fix lint errors

LFBarreto commented on package.json:
```diff
@@ -219,7 +220,7 @@
     "console": "console-browserify",
     "constants": "constants-browserify",
-    "crypto": "react-native-crypto",
+    "crypto": "crypto-js",
```
could you remove this from this PR it goes a bit out of the scope of the feature :+1: 
@valpinkman @MortalKastor to address the update needed and create a separate PR for it

tlaforge-ledger replied:
ok for me, do I need a Jira Ticket ? Which squad ? or only PR with @valpinkman and @MortalKastor mentionned ?

LFBarreto commented on src/components/CheckLanguageAvailability.js:
```diff
@@ -0,0 +1,104 @@
+                title={
+                  <>
+                    <Trans i18nKey="systemLanguageAvailable.switchButton" />
+                    <Text> {languages[osLanguage]}</Text>
```
you could maybe use the language in the i18n template in that case :thinking: 
```
<Trans i18nKey="systemLanguageAvailable.switchButton" values={{language: languages[osLanguage] }} />
... en.app.json
"switchButton":  "switch to {{language}}",
...

LFBarreto commented on src/context/Locale.js:
```diff
@@ -89,3 +89,34 @@ export default function LocaleProvider({ children }: Props) {
+  }, []);
+
+  useEffect(() => {
+    hasAnsweredLanguageAvailable().then(answerLanguageAvailable);
```
did you mean to 
 ```
 hasAnsweredLanguageAvailable().then(setChecked)
 ```
 instead :thinking:

LFBarreto reviewed (changes requested):
Some minor changes required :+1:

tlaforge-ledger committed:
use variable for switch button

tlaforge-ledger committed:
fix hook

gre commented on package.json:
```diff
@@ -94,6 +94,7 @@
     "color": "^3.1.3",
     "console-browserify": "^1.2.0",
     "constants-browserify": "^1.0.0",
+    "crypto-js": "3.1.9-1",
```
unclear why this is related to this PR.

LFBarreto committed:
(Languages): fix issue in discoverability modal

OlivierFreyssinet committed:
Fix typo

ofreyssinet-ledger commented:
Fixed the typo as requested in [LL-8921]

OlivierFreyssinet committed:
Default language + Refactor language setting and i18next initialization

- The settings.language field of the store is now the source of truth that gets fed to i18next
- settings.language is initialised with the system language if it is one of the supported languages, falling back to english

OlivierFreyssinet committed:
Use useTranslation rather than <Trans /> component on screen where language can change

That ensures the text gets updated when i18n.language changes.
https://react.i18next.com/latest/trans-component#important-note

ofreyssinet-ledger commented:
Just rebased and force pushed.
I added some commits with the main purpose of having the system's language used as a default for a new user in case that language is supported. [LL-7059]

I also did some refactoring regarding where is the source of truth for the language locale, it is now in the store's `setting.locale` field and i18next gets updated with that value when it changes.

This `setting.locale` field is the one being initialised to the system language if it's one of the supported languages.

I need this refactoring for another PR (https://github.com/LedgerHQ/ledger-live-mobile/pull/1941) regarding the new "region" distinct locale that we will have, as it needs a simple initialisation logic based on the language's initialisation logic, and it's a value that lives in the store, hence why I needed the logic of the language initialisation to be done there. I hope that makes sense, if not we can discuss it.

LFBarreto commented on src/components/CheckLanguageAvailability.js:
```diff
@@ -0,0 +1,111 @@
+    return null;
+  }
+
+  const onRequestClose = () => {
```
maybe use Callback here to not re-render these callbacks :thinking: 
the show or not show doesn't impact it so you can still do hooks and have a condition afterwards

ofreyssinet-ledger replied:
done https://github.com/LedgerHQ/ledger-live-mobile/pull/1926/commits/2b2ebb9aa33a2836d9e2181f28084af19b01debd

OlivierFreyssinet committed:
lint


Lint


pourquoi mon auto lint fonctionne pas :'(

OlivierFreyssinet committed:
Use useCallback to avoid unnecessary rerenders

OlivierFreyssinet committed:
Fix Locale.js

OlivierFreyssinet committed:
Remove legacy flow-defs/react-i18next.js (`yarn flow` passes)

Definitions are incomplete and the full ones can be retrieved from the lib itself.

OlivierFreyssinet committed:
Fix description in CheckLanguageAvailability modal

OlivierFreyssinet committed:
Fix flow

OlivierFreyssinet committed:
Merge branch 'develop' into LL-7025-Language-discoverability
